main.py

```diff
@@ -68,6 +70,12 @@ def setup(self):
         self._check_consistency_exec_at = {}
         self._check_consistency_locks = defaultdict(Lock)
 
+        self._pending_barrier_reply = defaultdict(dict)
```

Comment:
I'm not completely sure if we need to any potential action for barrier replies that never get replied, we probably need to periodically clean this up, any suggestions? Maybe we could run a clean up every 30 mins or so? This is also a sort of thing that we could have in the database and use a TTL to auto delete if we had more feature-rich DB.

Reply:
I'm not sure if I understood, but if the barrier request don't get replied it means that the switch didnt confirm the previous request so the next request should not be sent. Wdyt? It looks like the barrier req/repl are being handled in another execution path from the user request (not inline). If that is so, the only we can do with the unanswered barrier request is to ignore then. A clean up procedure would be nice.

Reply:
Right, @italovalcy, the request/reply are being handled in different threads asynchronously without the sending thread waiting for it, [sent here](https://github.com/kytos-ng/flow_manager/blob/feature/barrier/main.py#L699), and [received here](https://github.com/kytos-ng/flow_manager/blob/feature/barrier/main.py#L164-L165). 

I think we could try to queue and send subsequent flow mods only after a reply was either received or resulted in timeout, but that would serialize (in time) all of the requests that currently can be happen concurrently with some locks, and we also would need to fetch the flow mods from the DB in order accordingly. The reason why I didn't initially went down this route, was primarily because of this quote in the OF 1.3 spec:

> A.3.8 Barrier Message
    ... Upon receipt, the switch must finish processing all previously-received messages, including sending corresponding reply or error messages, **before executing any messages beyond the Barrier Request**. 

So, theoretically, since the messages will be sent in order here and queued out in the `msg_out` core `KytosBuffers` queue instance, I'd expect that even if subsequent flow mod and barriers end up being sent without getting a reply they'd still be processed correctly and queued in the switch since the switch isn't suppose to execute subsequent messages, so facilitating our implementation and code maintenance here. Let me know if you share the same understanding.

Reply:
@italovalcy regarding the upper bound limit, cleaning up, I've mapped it on this issue https://github.com/kytos-ng/flow_manager/issues/57

Reply:
> Right, @italovalcy, the request/reply are being handled in different threads asynchronously without the sending thread waiting for it, [sent here](https://github.com/kytos-ng/flow_manager/blob/feature/barrier/main.py#L699), and [received here](https://github.com/kytos-ng/flow_manager/blob/feature/barrier/main.py#L164-L165).
> 
> I think we could try to queue and send subsequent flow mods only after a reply was either received or resulted in timeout, but that would serialize (in time) all of the requests that currently can be happen concurrently with some locks, and we also would need to fetch the flow mods from the DB in order accordingly. The reason why I didn't initially went down this route, was primarily because of this quote in the OF 1.3 spec:
> 
> > A.3.8 Barrier Message
> > ... Upon receipt, the switch must finish processing all previously-received messages, including sending corresponding reply or error messages, **before executing any messages beyond the Barrier Request**.
> 
> So, theoretically, since the messages will be sent in order here and queued out in the `msg_out` core `KytosBuffers` queue instance, I'd expect that even if subsequent flow mod and barriers end up being sent without getting a reply they'd still be processed correctly and queued in the switch since the switch isn't suppose to execute subsequent messages, so facilitating our implementation and code maintenance here. Let me know if you share the same understanding.

Understood! thanks for the clarification, @viniarck